sync-android/src/test/java/com/cloudant/sync/datastore/encryption/CachingKeyProviderTests.java

```diff
@@ -0,0 +1,131 @@
+package com.cloudant.sync.datastore.encryption;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.junit.Assume.assumeNotNull;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Arrays;
+
+public class CachingKeyProviderTests {
```

Comment:
I don't think you need to use AndroidKeyProvider for these tests. I think this will just slow down the test through the key generation and cleanup. I'd suggest running most of the tests with just the SimpleKeyProvider. Then you control the key used, so you can make sure the right key is returned (avoiding the checks that the keys are just "different" or the "same" without checking their values against known good versions).

In addition, you should write a test to make sure the key is actually being cached. I'd do this by creating an implementation of KeyProvider which returns a different key each time its getEncryptionKey method is called (e.g., use Random() to get a new key each time). When wrapped with CachingKeyProvider, if the key is being cached correctly, the same key will be returned by the CachingKeyProvider whereas if the key isn't being cached correctly a different key will be returned each time from the underlying `RandomKeyProvider`.]

I.e.:

```java
private class RandomKeyProvider implements KeyProvider {
  public EncryptionKey getEncryptionKey() {
    byte[] buf = new byte[32];
    new Random().nextBytes(buf);
    return new EncryptionKey(buf);
  }
}
```